pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,680 @@
+			Kind:    apiDes.Kind,
+		}
+
+		if _, ok := s.apiServiceDescriptionMap[apiGVK]; ok {
+			return fmt.Errorf("APIService Descriptors with duplicated GVK %v are found", apiGVK)
+		}
+		api := apiDes
+		s.apiServiceDescriptionMap[apiGVK] = &api
+	}
+
+	for _, crdDes := range s.registryV2CSV.CustomResourceDefinitions {
+		parts := strings.SplitN(crdDes.Name, ".", 2)
+		if len(parts) != 2 {
+			return fmt.Errorf("couldn't parse plural.group from crd name: %s", crdDes.Name)
+		}
+		crdGVKN := v1.GroupVersionKind{
+			Group:   parts[1],
+			Version: crdDes.Version,
+			Kind:    crdDes.Kind,
+		}
+
+		if _, ok := s.customResourceDefinitionMap[crdGVKN]; ok {
+			return fmt.Errorf("CustomResourceDefinitions Descriptors with duplicated GVK %v are found", crdGVKN)
+		}
+		crd := crdDes
+		s.customResourceDefinitionMap[crdGVKN] = &crd
+	}
+	return nil
+}
+
+// appendTostructuredObjectMap converts the unstructured object and try to convert it into a specified interface
+// before appending to the structuredObjectMap. It only returns error if the unstructured is not convertible.
+func (s *synthesize) appendTostructuredObjectMap(kind string, bufObj unstructured.Unstructured) {
+	if existingObj, ok := s.structuredObjectMap[kind]; !ok {
+		s.structuredObjectMap[kind] = &[]unstructured.Unstructured{bufObj}
+	} else {
+		*existingObj = append(*existingObj, bufObj)
+	}
+}
+
+// loadDependencies tries to decode an object into registry.DependenciesFile{}.
+// It updates the dependencies in synthesize once decode is successful.
+func (s *synthesize) loadDependencies(path string) error {
+	dependenciesFile := registry.DependenciesFile{}
+	if err := registry.ParseDependenciesFile(path, &dependenciesFile); err != nil {
+		return err
+	}
+
+	s.dependencies = &dependenciesFile
+	return nil
+}
+
+// getRegistryV2CSV gets the RegistryV2CSV object ensures that is loaded.
+func (s *synthesize) getRegistryV2CSV() (*RegistryV2CSV, error) {
+	if s.registryV2CSV != nil {
+		return s.registryV2CSV, nil
+	}
+	return nil, fmt.Errorf("the %s resource is not found in bundle", RegistryV2CSVKind)
+}
+
+// getInstallStrategy converts all Deployment kind object in bundle into install strategy and extracts the
+// service account name referred in the deployments. It then adds all roleRefs from (cluster)RoleBindings that mentions
+// any of the service accounts, adding the rules of (cluster)roles paring with their correspondent ServceAccounts into
+// (cluster)permissions.
+func (s *synthesize) getInstallStrategy() (*v1alpha1.NamedInstallStrategy, error) {
+	deploys, err := s.getDeploytments()
+	if err != nil {
+		return nil, fmt.Errorf("deployments are required for synthesizing CSV, %v", err)
+	}
+
+	var deploymentSpecs []v1alpha1.StrategyDeploymentSpec
+	deploySANames := make(map[string]struct{})
+
+	for _, deploy := range *deploys {
+		deployName := deploy.Name
+		if deployName == "" {
+			deployName = names.SimpleNameGenerator.GenerateName(deploy.GetGenerateName())
```

Comment:
We are getting Kube Deployment resources as install strategies. This would allow us to process Deployment resource that has `GenerateName` instead of Name. 

If we just care about testing, we can use seed in test.

Reply:
> If we just care about testing, we can use seed in test.
Good idea! If we want to support this feature, then we should probably always use a seed (think `make verify` CI situations).

Though, this won't lead to the same behavior at runtime as a user putting a generate name on a deployment might expect; i.e. a different name for each application.